coco/scheduler.py

```diff
@@ -3,3 +3,129 @@
+
+    async def start(self):
+        """Start the scheduler (async)."""
+        self.tasks = []
+        for timer in self.timers:
+            logger.debug(f"Setting timer '{timer.name}' every {timer.period} s.")
+            task = asyncio.create_task(timer.run())
+            self.tasks.append(task)
+        await asyncio.gather(*self.tasks)
+
+    def stop(self):
+        """Stop the scheduler. Cancels all timer tasks."""
+        for task in self.tasks:
+            task.cancel()
+
+    def _gen_timers(self, endpoints):
+        if len(self.timers) > 0:
+            raise Exception("Timers already exist.")
+
+        for edpt in endpoints.values():
+            if edpt.schedule is not None:
+                if edpt.values is not None:
+                    raise ValueError(
+                        f"Endpoint '{edpt.name}' cannot be scheduled with a 'values' config block."
+                    )
+                try:
+                    period = edpt.schedule["period"]
+                except KeyError:
+                    raise ValueError(
+                        f"Endpoint '{edpt.name}' schedule block must include 'period'."
+                    )
+                period = str2total_seconds(period)
+                if period is None or period == 0:
+                    raise ValueError(
+                        f"Could not parse 'period' parameter for endpoint {edpt.name}"
+                    )
+                timer = EndpointTimer(period, edpt, self.host, self.port)
+                self.timers.append(timer)
+
+
+class Timer(object):
+    """Asynchronous timer."""
+
+    def __init__(self, name, period):
+        self.name = name
+        self.period = period
+        self._last_t = time()
+        self._start_t = self._last_t
+
+    async def run(self):
+        """Start the timer (async)."""
+        while True:
+            t = self._wait_time()
+            if t > 0:
+                try:
+                    await asyncio.sleep(t)
+                except asyncio.CancelledError:
+                    logger.debug(f"Cancelled timer '{self.name}''.")
+                    break
+            self._last_t = time()
+            await self._call()
+
+    async def _call(self):
+        """Override this method with whatever the timer does."""
+        pass
+
+    def _wait_time(self):
+        """Time to wait until next execution."""
+        return self.period - (time() - self._last_t)
+
+
+class EndpointTimer(Timer):
+    """Timer that calls a coco endopint."""
+
+    def __init__(self, period, endpoint, host, port):
+        self.endpoint = endpoint
+        self.host, self.port = host, port
+        super().__init__(endpoint.name, period)
+
+    async def _call(self):
+        # logger.debug(f"{self.name}: {time() - self._start_t}")
+
+        # Send request to coco
+        url = f"http://{self.host}:{self.port}/{self.name}"
+        async with request(self.endpoint.type, url) as r:
+            r.raise_for_status()
```

Comment:
Where is this caught?

Reply:
It doesn't get caught at the moment. What would you like it to do if this exception is caught?

Reply:
I think it would be best to catch `BaseException`'s from these lines and then write a `logger.error` line like "Scheduler failed calling {endpoint}. Has coco's sanic server crashed?" followed by exiting coco.

Reply:
and `: {e}`

Reply:
OK, I have that in now, but since the scheduler runs in a thread it can't exit coco itself. We could have it post a slack warning instead.

Reply:
Yeah, the slack messages we plan to hook up with the logging. So `error` will always go to slack. But also we have to synchronize exiting between all threads and processes. We can leave it like that in this PR, though, I think.